drivers/gpu/drm/panel/panel-samsung-s6e3fa0-full.c

```diff
@@ -0,0 +1,764 @@
+				      0x00, 0x00, 0x00, 0x00, 0x03, 0x55, 0x54,
+				      0x20, 0x00, 0x0a, 0xaa, 0xaf, 0x0f, 0x02,
+				      0x22, 0x22, 0x10);
+
+		/* Unknown command: 16 frame Averaging (0x41) */
+		dsi_generic_write_seq(dsi, 0xb5, 0x41);
+
+		s6e3fa0_write_gamma_apply(ctx);
+
+		/* TE Vsync ON  */
+		dsi_generic_write_seq(dsi, 0xb0, 0x02);
+		dsi_generic_write_seq(dsi, 0xfd, 0x0a);
+		dsi_generic_write_seq(dsi, 0xfe, 0x80);
+		dsi_generic_write_seq(dsi, 0xfe, 0x00);
+
+		mipi_dsi_dcs_set_tear_on(dsi, MIPI_DSI_DCS_TEAR_MODE_VBLANK);
+
+		/* Touch Hsync Enable */
+		dsi_generic_write_seq(dsi, 0xbd, 0x05, 0x02, 0x02);
+	} else {
+		dev_warn(&dsi->dev, "init: skipped, unknown panel!\n");
+	}
+
+	s6e3fa0_test_key_toggle(ctx, false, TK_LEVEL_2);
+
+	return 0;
+}
+
+static int s6e3fa0_power_on(struct s6e3fa0_ctx *ctx)
+{
+	int ret;
+
+	ret = regulator_bulk_enable(ARRAY_SIZE(ctx->supplies), ctx->supplies);
+	if (ret < 0) {
+		dev_err(&ctx->dsi->dev, "Failed to enable regulators: %d\n",
+			ret);
+		return ret;
+	}
+	msleep(30);
+
+	/* send reset pulse */
+	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+	usleep_range(5000, 6000);
+	gpiod_set_value_cansleep(ctx->reset_gpio, 0);
+	usleep_range(7000, 8000);
+
+	/* Panel init sequence does not work without this delay! */
+	msleep(60);
+
+	return 0;
+}
+
+static int s6e3fa0_power_off(struct s6e3fa0_ctx *ctx)
+{
+	int ret;
+
+	gpiod_set_value(ctx->reset_gpio, 1);
+
+	ret = regulator_bulk_disable(ARRAY_SIZE(ctx->supplies), ctx->supplies);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static int s6e3fa0_enable(struct drm_panel *panel)
+{
+	struct s6e3fa0_ctx *ctx = panel_to_s6e3fa0(panel);
+	struct mipi_dsi_device *dsi = ctx->dsi;
+	int ret;
+
+	ret = mipi_dsi_dcs_exit_sleep_mode(dsi);
+	if (ret < 0) {
+		dev_err(&dsi->dev, "Failed to exit sleep mode: %d\n", ret);
+		return ret;
+	};
+	msleep(60);
+
+	ret = mipi_dsi_dcs_set_display_on(dsi);
+	if (ret < 0) {
+		dev_err(&dsi->dev, "Failed to set display on: %d\n", ret);
+		return ret;
+	}
+	msleep(50);
+
+	return 0;
+}
+
+static int s6e3fa0_disable(struct drm_panel *panel)
+{
+	struct s6e3fa0_ctx *ctx = panel_to_s6e3fa0(panel);
+	struct mipi_dsi_device *dsi = ctx->dsi;
+	int ret;
+
+	dsi->mode_flags &= ~MIPI_DSI_MODE_LPM;
+
+	ret = mipi_dsi_dcs_set_display_off(dsi);
+	if (ret < 0) {
+		dev_err(&dsi->dev, "Failed to set display off: %d\n", ret);
+		return ret;
+	}
+	usleep_range(10000, 11000);
+
+	ret = mipi_dsi_dcs_enter_sleep_mode(dsi);
+	if (ret < 0) {
+		dev_err(&dsi->dev, "Failed to enter sleep mode: %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int s6e3fa0_prepare(struct drm_panel *panel)
+{
+	struct s6e3fa0_ctx *ctx = panel_to_s6e3fa0(panel);
+	struct device *dev = &ctx->dsi->dev;
+	int ret;
+
+	ret = s6e3fa0_power_on(ctx);
+	if (ret < 0) {
+		dev_err(dev, "Failed to power on the panel!\n");
+		return ret;
+	}
+
+	/* Detect panel subtype */
+	ret = s6e3fa0_detect_panel_type(ctx);
+	if (ret < 0)
+		dev_warn(dev, "Failed to detect LCD type!\n");
```

Comment:
if you keep s6e3fa0_detect_panel_type(), then this should probably return failure here, and not continue to send init sequence

Reply:
BTW I hope init seq is not just copy from s6e3fa2 driver? Have you verified your init sequence to your downstream panel dtsi? Can you power off the display and turn it back again okay?

Reply:
Returned failure if panel ID is wrong and don't send init sequence

Reply:
Updated init sequence, verified it's still working.